### DISCUSSION TITLE
Upsert with a nested to-many set

### DISCUSSION THREAD

**Discussion by pm-dev** [prisma/prisma#22501] [Q&A]
### Question

Take a list with items:
```prisma
model TodoList {
  id               Int          @id @default(autoincrement())
  title           String
  authorId    Int 
  items        TodoItem[]
  @unique([authorId, title])
}
model TodoItem {
    id        Int          @id @default(autoincrement())
   name  String 
   listId   String
   list      TodoList  @relation(fields: [listId], references: [title])
   @unique([listId, name])
}
```
You can imagine wanting to build an API where a client wants to save a list with items. If the list already exists, it's overwritten. Is it possible to do this in prisma with one query?
```js
prisma.todoList.upsert({
      where: {
        authorId_title: { title: args.title, authorId: args.authorId }
      },
      create: {
        title: args.title,
        items: {
          createMany: {
            data: args.items.map((item) => {
              return {
                name: item.name
              }
            })
          }
        }
      },      
      update: {
        items: {
            set: args.items.map((item) => {
                return {
                    listId_name: { listId: ** I don't know the list Id**,  name: item.name },
                    name: item.name
                }
            })
          }
        }
      }
```
In the update -> set, I should be able to identify the list via `authorId_title`, since I don't know the `listId`, or even better, not at all since it's nested within the list's own upsert.

Am I missing anything or is this not possible at the moment?

### Information about Prisma Schema, Client Queries and Environment (optional)

On prisma: 5.7.1

**Comment by ludralph** [comment]
Hi @pm-dev 👋 

Thank you for raising this question.

Prisma Client does not currently support upsert operations on related entries with composite IDs in a single query when the ID is not known. This can be seen in this open [feature request](https://github.com/prisma/prisma/issues/8324).

It is not possible to perform such an operation in a single query and from the issue there was a suggestion to querying the document first for the ID as a workaround while also using interactive transactions to ensure the operations are executed within a single transactional context.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by pm-dev** [reply]
@ludralph Thanks for the reply.

The fix from the perspective of prisma client's interface seems simple enough. Wherever we have a `CompoundUniqueInput` if one of the compound elements is an `@id` from the prisma schema, instead of only being able to provide `fooId: number`, you could instead provide: `foo: FooWhereUniqueInput`

Not sure how hard this would be to implement and what performance impact it might have, but this seems like a reasonable, non-breaking, update to the clients API.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
